File: annotations-runtime/src/main/java/com/github/dynamicextensionsalfresco/webscripts/AnnotationWebScriptBuilder.java
package com.github.dynamicextensionsalfresco.webscripts;

import com.github.dynamicextensionsalfresco.webscripts.annotations.Attribute;
import com.github.dynamicextensionsalfresco.webscripts.annotations.Authentication;
import com.github.dynamicextensionsalfresco.webscripts.annotations.Before;
import com.github.dynamicextensionsalfresco.webscripts.annotations.Cache;
import com.github.dynamicextensionsalfresco.webscripts.annotations.ExceptionHandler;
import com.github.dynamicextensionsalfresco.webscripts.annotations.HttpMethod;
import com.github.dynamicextensionsalfresco.webscripts.annotations.Transaction;
import com.github.dynamicextensionsalfresco.webscripts.annotations.Uri;
import com.github.dynamicextensionsalfresco.webscripts.annotations.WebScript;
import com.github.dynamicextensionsalfresco.webscripts.arguments.HandlerMethodArgumentsResolver;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.beans.factory.BeanFactoryAware;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
import org.springframework.core.annotation.AnnotationUtils;
import org.springframework.extensions.webscripts.Description;
import org.springframework.extensions.webscripts.Description.FormatStyle;
import org.springframework.extensions.webscripts.Description.Lifecycle;
import org.springframework.extensions.webscripts.Description.RequiredAuthentication;
import org.springframework.extensions.webscripts.Description.RequiredTransaction;
import org.springframework.extensions.webscripts.Description.TransactionCapability;
import org.springframework.extensions.webscripts.DescriptionImpl;
import org.springframework.extensions.webscripts.TransactionParameters;
import org.springframework.util.Assert;
import org.springframework.util.ClassUtils;
import org.springframework.util.ReflectionUtils;
import org.springframework.util.StringUtils;

/**
 * Creates {@link AnnotationWebScript} instances from beans defined in a {@link BeanFactory}.
 *
 * @author Laurens Fridael
 * @author Laurent Van der Linden
 */
public final class AnnotationWebScriptBuilder implements BeanFactoryAware {

    @Autowired
    private HandlerMethodArgumentsResolver handlerMethodArgumentsResolver;
    @Autowired
    private MessageConverterRegistry messageConverterRegistry;

    public final void setHandlerMethodArgumentsResolver(@NotNull HandlerMethodArgumentsResolver value) {
        if (value == null) {
            throw new IllegalArgumentException("value is null");
        }
        this.handlerMethodArgumentsResolver = value;
    }

    public final void setMessageConverterRegistry(@NotNull MessageConverterRegistry value) {
        if (value == null) {
            throw new IllegalArgumentException("value is null");
        }
        this.messageConverterRegistry = value;
    }

    /* Dependencies */

    @Nullable
    private ConfigurableListableBeanFactory beanFactory;
    private final String trailingSlashExpression = "/$";
    private final String leadingSlashExpression = "^/";

    @Nullable
    protected final ConfigurableListableBeanFactory getBeanFactory() {
        return this.beanFactory;
    }

    /* Main operations */

    /**
     * Creates {@link AnnotationWebScript}s from a given named bean by scanning methods annotated with {@link Uri}.
     *
     * @return The {@link AnnotationWebScript} or null if the implementation does not consider the bean to be a handler
     * *         for an {@link AnnotationWebScript}.
     */
    @NotNull
    public final List<org.springframework.extensions.webscripts.WebScript> createWebScripts(
            @NotNull final String beanName) {
        Assert.hasText(beanName, "Bean name cannot be empty.");

        if (beanFactory == null) {
            throw new IllegalStateException("beanFactory");
        }
        Class<?> beanType = beanFactory.getType(beanName);
        if (beanType == null) {
            return Collections.emptyList();
        }
        final WebScript webScriptAnnotation =
                beanFactory.findAnnotationOnBean(beanName, WebScript.class) != null
                        ? beanFactory.findAnnotationOnBean(beanName, WebScript.class)
                        : getDefaultWebScriptAnnotation();
        String baseUri = webScriptAnnotation.baseUri();
        if (StringUtils.hasText(baseUri) && !baseUri.startsWith("/")) {
            throw new RuntimeException("@WebScript baseUri for class '" + beanType + "' does not start with a slash: '"
                    + baseUri + "'");
        }

        final HandlerMethods handlerMethods = new HandlerMethods();
        final ArrayList<org.springframework.extensions.webscripts.WebScript> webScripts = new ArrayList<>();
        final MethodHandler methodHandler = new MethodHandler(beanName, handlerMethods, webScriptAnnotation,
                webScripts);
        ReflectionUtils.doWithMethods(beanType, methodHandler::registerBeforeHandlerMethod);
        ReflectionUtils.doWithMethods(beanType, methodHandler::registerAttributeHandlerMethod);
        ReflectionUtils.doWithMethods(beanType, methodHandler::registerExceptionHandlerMethod);
        ReflectionUtils.doWithMethods(beanType, methodHandler::registerWebscriptHandlerMethod);

        HashSet<String> ids = new HashSet<>();
        for (org.springframework.extensions.webscripts.WebScript webScript : webScripts) {
            final String webscriptId = webScript.getDescription().getId();
            final boolean notContained = ids.add(webscriptId);
            if (!notContained) {
                throw new IllegalStateException("Duplicate Web Script ID \"" + webscriptId
                        + "\" Make sure handler methods of annotation-based Web Scripts have unique names.");
            }
        }

        return webScripts;
    }

    private class MethodHandler {

        private final String beanName;
        private final HandlerMethods handlerMethods;
        private final WebScript webScriptAnnotation;
        private final List<org.springframework.extensions.webscripts.WebScript> webScripts;

        public MethodHandler(String beanName, HandlerMethods handlerMethods,
                WebScript webScriptAnnotation, List<org.springframework.extensions.webscripts.WebScript> webScripts) {
            this.beanName = beanName;
            this.handlerMethods = handlerMethods;
            this.webScriptAnnotation = webScriptAnnotation;
            this.webScripts = webScripts;
        }

        private void registerBeforeHandlerMethod(Method method) {
            if (method == null) {
                throw new IllegalArgumentException("method is null");
            }
            Before before = AnnotationUtils.findAnnotation(method, Before.class);
            if (before != null) {
                if (AnnotationUtils.findAnnotation(method, Attribute.class) != null
                        || AnnotationUtils.findAnnotation(method, Uri.class) != null) {
                    throw new RuntimeException(
                            "Cannot combine @Before, @Attribute and @Uri on a single method. Method: "
                                    + ClassUtils.getQualifiedMethodName(method));
                }
                handlerMethods.getBeforeMethods().add(method);
            }
        }

        private void registerAttributeHandlerMethod(Method method) {
            if (method == null) {
                throw new IllegalArgumentException("method is null");
            }
            Attribute attribute = AnnotationUtils.findAnnotation(method, Attribute.class);
            if (attribute != null) {
                if (AnnotationUtils.findAnnotation(method, Before.class) != null
                        || AnnotationUtils.findAnnotation(method, Uri.class) != null) {
                    throw new RuntimeException(
                            "Cannot combine @Before, @Attribute and @Uri on a single method. Method: "
                                    + ClassUtils.getQualifiedMethodName(method));
                }
                if (Void.TYPE.equals(method.getReturnType())) {
                    throw new RuntimeException("@Attribute methods cannot have a void return type.");
                }
                handlerMethods.getAttributeMethods().add(method);
            }
        }

        private void registerExceptionHandlerMethod(Method method) {
            if (method == null) {
                throw new IllegalArgumentException("method is null");
            }
            ExceptionHandler exceptionHandler = AnnotationUtils.findAnnotation(method, ExceptionHandler.class);
            if (exceptionHandler != null) {
                if (AnnotationUtils.findAnnotation(method, Attribute.class) != null
                        || AnnotationUtils.findAnnotation(method, Before.class) != null
                        || AnnotationUtils.findAnnotation(method, Uri.class) != null) {
                    throw new RuntimeException(
                            "Cannot combine @Before, @Attribute @ExceptionHandler or @Uri on a single method. Method: "
                                    + ClassUtils.getQualifiedMethodName(method));
                }
                handlerMethods.getExceptionHandlerMethods().add(new ExceptionHandlerMethod(exceptionHandler, method));
            }
        }

        private void registerWebscriptHandlerMethod(Method method) {
            if (method == null) {
                throw new IllegalArgumentException("method is null");
            }
            Uri uri = AnnotationUtils.findAnnotation(method, Uri.class);
            if (uri != null) {
                org.springframework.extensions.webscripts.WebScript webScript =
                        createWebScript(beanName, webScriptAnnotation, uri, handlerMethods.createForUriMethod(method), uri.method().name());
                webScripts.add(webScript);
                if(uri.method() == HttpMethod.GET) {
                    org.springframework.extensions.webscripts.WebScript headWebscript =
                            createWebScript(beanName, webScriptAnnotation, uri, handlerMethods.createForUriMethod(method), "HEAD");
                    webScripts.add(headWebscript);
                }
            }
        }
    }

    /* Utility operations */

    @NotNull
    protected final AnnotationWebScript createWebScript(@NotNull String beanName, @NotNull WebScript webScript,
            @NotNull Uri uri, @NotNull HandlerMethods handlerMethods, String httpMethod) {
        if (beanName == null) {
            throw new IllegalArgumentException("beanName is null");
        }
        if (webScript == null) {
            throw new IllegalArgumentException("webScript is null");
        }
        if (uri == null) {
            throw new IllegalArgumentException("uri is null");
        }
        if (handlerMethods == null) {
            throw new IllegalArgumentException("handlerMethods is null");
        }
        DescriptionImpl description = new DescriptionImpl();
        if (StringUtils.hasText(webScript.defaultFormat())) {
            description.setDefaultFormat(webScript.defaultFormat());
        }
        final String baseUri = webScript.baseUri();
        this.handleHandlerMethodAnnotation(uri, handlerMethods.getUriMethod(), description, baseUri, httpMethod);
        this.handleTypeAnnotations(beanName, webScript, description);
        String id = String.format("%s.%s.%s", generateId(beanName), handlerMethods.getUriMethod().getName(),
                httpMethod.toLowerCase());
        description.setId(id);

        if (beanFactory == null) {
            throw new IllegalStateException("beanFactory");
        }
        Object handler = beanFactory.getBean(beanName);
        description.setStore(new DummyStore());
        return createWebScript(description, handler, handlerMethods);
    }

    @NotNull
    protected final AnnotationWebScript createWebScript(@NotNull Description description, @NotNull Object handler,
            @NotNull HandlerMethods handlerMethods) {
        if (description == null) {
            throw new IllegalArgumentException("description is null");
        }
        if (handler == null) {
            throw new IllegalArgumentException("handler is null");
        }
        if (handlerMethods == null) {
            throw new IllegalArgumentException("handlerMethods is null");
        }
        return new AnnotationWebScript(description, handler, handlerMethods, this.handlerMethodArgumentsResolver,
                this.messageConverterRegistry);
    }

    protected final void handleHandlerMethodAnnotation(@NotNull Uri uri, @NotNull Method method,
            @NotNull DescriptionImpl description, @NotNull String baseUri, String httpMethod) {
        Assert.notNull(uri, "Uri cannot be null.");
        Assert.notNull(method, "HttpMethod cannot be null.");
        Assert.notNull(description, "Description cannot be null.");
        if (baseUri == null) {
            throw new IllegalArgumentException("baseUri is null");
        }

        String[] uris;
        if (uri.value().length > 0) {
            uris = Arrays.stream(uri.value())
                    .map(it -> baseUri.replaceAll(trailingSlashExpression, "") + "/" +
                            it.replaceAll(leadingSlashExpression, ""))
                    .toArray(String[]::new);
        } else if (StringUtils.hasText(baseUri)) {
            uris = new String[]{baseUri.replaceAll(trailingSlashExpression, "")};
        } else {
            throw new RuntimeException(
                    String.format(
                            "No value specified for @Uri on method '%s' and no base URI found for @WebScript on class."
                            , ClassUtils.getQualifiedMethodName(method))
            );
        }
        description.setUris(uris);
        /*
         * For the sake of consistency we translate the HTTP method from the HttpMethod enum. This also shields us from
         * changes in the HttpMethod enum names.
         */
        description.setMethod(httpMethod);
        /*
         * Idem dito for FormatStyle.
         */
        Description.FormatStyle springFormatStyle;
        switch (uri.formatStyle()) {
            case ANY:
                springFormatStyle = FormatStyle.any;
                break;
            case ARGUMENT:
                springFormatStyle = FormatStyle.argument;
                break;
            case EXTENSION:
                springFormatStyle = FormatStyle.extension;
                break;
            default:
                throw new IllegalArgumentException("Unknown format style: " + uri.formatStyle());
        }
        description.setFormatStyle(springFormatStyle);
        if (StringUtils.hasText(uri.defaultFormat())) {
            description.setDefaultFormat(uri.defaultFormat());
        }
        description.setMultipartProcessing(uri.multipartProcessing());

        Authentication methodAuthentication = method.getAnnotation(Authentication.class);
        if (methodAuthentication != null) {
            this.handleAuthenticationAnnotation(methodAuthentication, description);
        }

        Transaction methodTransaction = method.getAnnotation(Transaction.class);
        if (methodTransaction != null) {
            this.handleTransactionAnnotation(methodTransaction, description);
        }
    }

    protected final void handleTypeAnnotations(@NotNull String beanName, @NotNull WebScript webScript,
            @NotNull DescriptionImpl description) {
        if (beanName == null) {
            throw new IllegalArgumentException("beanName is null");
        }
        if (webScript == null) {
            throw new IllegalArgumentException("webScript is null");
        }
        if (description == null) {
            throw new IllegalArgumentException("description is null");
        }

        this.handleWebScriptAnnotation(webScript, beanName, description);

        if (this.beanFactory == null) {
            throw new IllegalStateException("beanFactory");
        }

        if (description.getRequiredAuthentication() == null) {
            Authentication authentication = beanFactory.findAnnotationOnBean(beanName, Authentication.class);
            if (authentication == null) {
                authentication = this.getDefaultAuthenticationAnnotation();
            }
            this.handleAuthenticationAnnotation(authentication, description);
        }

        if (description.getRequiredTransactionParameters() == null) {
            Transaction transaction = beanFactory.findAnnotationOnBean(beanName, Transaction.class);
            if (transaction == null) {
                if (description.getMethod().equals("GET")) {
                    transaction = this.getDefaultReadonlyTransactionAnnotation();
                } else {
                    transaction = this.getDefaultReadWriteTransactionAnnotation();
                }
            }

            this.handleTransactionAnnotation(transaction, description);
        }

        Cache cache = beanFactory.findAnnotationOnBean(beanName, Cache.class);
        if (cache == null) {
            cache = this.getDefaultCacheAnnotation();
        }
        this.handleCacheAnnotation(cache, beanName, description);

        description.setDescPath("");
    }

    protected final void handleWebScriptAnnotation(@NotNull WebScript webScript, @NotNull String beanName,
            @NotNull DescriptionImpl description) {
        Assert.notNull(webScript, "Annotation cannot be null.");
        Assert.hasText(beanName, "Bean name cannot be empty.");
        Assert.notNull(description, "Description cannot be null.");
        Assert.hasText(description.getMethod(), "Description method is not specified.");

        if (StringUtils.hasText(webScript.value())) {
            description.setShortName(webScript.value());
        } else {
            description.setShortName(this.generateShortName(beanName));
        }
        if (StringUtils.hasText(webScript.description())) {
            description.setDescription(webScript.description());
        } else {
            if (beanFactory == null) {
                throw new IllegalStateException("beanFactory");
            }
            description.setDescription(
                    String.format("Annotation-based WebScript for class %s", beanFactory.getType(beanName).getName()));
        }
        if (webScript.families().length > 0) {
            description.setFamilys(new LinkedHashSet<>(Arrays.asList(webScript.families())));
        }
        Lifecycle springLifeCycle;
        switch (webScript.lifecycle()) {
            case NONE:
                springLifeCycle = Lifecycle.none;
                break;
            case DRAFT:
                springLifeCycle = Lifecycle.draft;
                break;
            case DRAFT_PUBLIC_API:
                springLifeCycle = Lifecycle.draft_public_api;
                break;
            case DEPRECATED:
                springLifeCycle = Lifecycle.deprecated;
                break;
            case INTERNAL:
                springLifeCycle = Lifecycle.internal;
                break;
            case PUBLIC_API:
                springLifeCycle = Lifecycle.public_api;
                break;
            case SAMPLE:
                springLifeCycle = Lifecycle.sample;
                break;
            default:
                throw new IllegalArgumentException("Unknown lifeCycle '" + webScript.lifecycle() + "'");
        }
        description.setLifecycle(springLifeCycle);
    }

    protected final void handleAuthenticationAnnotation(@NotNull Authentication authentication,
            @NotNull DescriptionImpl description) {
        Assert.notNull(authentication, "Annotation cannot be null.");
        Assert.notNull(description, "Description cannot be null.");

        if (StringUtils.hasText(authentication.runAs())) {
            description.setRunAs(authentication.runAs());
        }
        RequiredAuthentication requiredAuthentication;
        switch (authentication.value()) {
            case NONE:
                requiredAuthentication = RequiredAuthentication.none;
                break;
            case GUEST:
                requiredAuthentication = RequiredAuthentication.guest;
                break;
            case USER:
                requiredAuthentication = RequiredAuthentication.user;
                break;
            case ADMIN:
                requiredAuthentication = RequiredAuthentication.admin;
                break;
            default:
                throw new IllegalArgumentException("Unknown authentication value '" + authentication.value() + "'");
        }
        description.setRequiredAuthentication(requiredAuthentication);
    }

    protected final void handleTransactionAnnotation(@NotNull Transaction transaction,
            @NotNull DescriptionImpl description) {
        Assert.notNull(transaction, "Annotation cannot be null.");
        Assert.notNull(description, "Description cannot be null.");

        TransactionParameters transactionParameters = new TransactionParameters();
        RequiredTransaction requiredTransaction;
        switch (transaction.value()) {
            case NONE:
                requiredTransaction = RequiredTransaction.none;
                break;
            case REQUIRED:
                requiredTransaction = RequiredTransaction.required;
                break;
            case REQUIRES_NEW:
                requiredTransaction = RequiredTransaction.requiresnew;
                break;
            default:
                throw new IllegalArgumentException("Unknown transaction value '" + transaction.value() + "'");
        }
        transactionParameters.setRequired(requiredTransaction);
        if (transaction.readOnly()) {
            transactionParameters.setCapability(TransactionCapability.readonly);
        } else {
            transactionParameters.setCapability(TransactionCapability.readwrite);
        }
        transactionParameters.setBufferSize(transaction.bufferSize());
        description.setRequiredTransactionParameters(transactionParameters);
    }

    protected final void handleCacheAnnotation(@NotNull Cache cache, @NotNull String beanName,
            @NotNull DescriptionImpl description) {
        Assert.notNull(cache, "Annotation cannot be null.");
        Assert.hasText(beanName, "Bean name cannot be empty.");
        Assert.notNull(description, "Description cannot be null.");

        org.springframework.extensions.webscripts.Cache requiredCache = new org.springframework.extensions.webscripts.Cache();
        requiredCache.setNeverCache(cache.neverCache());
        requiredCache.setIsPublic(cache.isPublic());
        requiredCache.setMustRevalidate(cache.mustRevalidate());
        description.setRequiredCache(requiredCache);
    }

    @NotNull
    protected final String generateId(@NotNull String beanName) {
        Assert.hasText(beanName, "Bean name cannot be empty");

        if (beanFactory == null) {
            throw new IllegalStateException("beanFactory");
        }
        Class<?> clazz = beanFactory.getType(beanName);
        if (clazz == null) {
            throw new IllegalStateException("clazz is null");
        }
        String id = clazz.getName();
        if (id == null) {
            throw new IllegalStateException("clazz.name is null");
        }
        return id;
    }

    @NotNull
    protected final String generateShortName(@NotNull String beanName) {
        Assert.hasText(beanName, "Bean name cannot be empty");

        if (beanFactory == null) {
            throw new IllegalStateException("beanFactory");
        }
        final Class<?> clazz = beanFactory.getType(beanName);
        if (clazz == null) {
            throw new IllegalStateException("clazz is null");
        }
        final String shortName = ClassUtils.getShortName(clazz);
        if (shortName == null) {
            throw new IllegalStateException("ClassUtils.getShortName(clazz) is null");
        }
        return shortName;
    }

    /*
     * These methods use local classes to obtain annotations with default settings.
     */

    private Authentication getDefaultAuthenticationAnnotation() {
        @Authentication
        final class Default {

        }
        return Default.class.getAnnotation(Authentication.class);
    }

    private final Transaction getDefaultReadWriteTransactionAnnotation() {
        @Transaction
        final class Default {

        }
        return Default.class.getAnnotation(Transaction.class);
    }

    private Transaction getDefaultReadonlyTransactionAnnotation() {
        @Transaction(readOnly = true)
        final class Default {

        }
        return Default.class.getAnnotation(Transaction.class);
    }

    private Cache getDefaultCacheAnnotation() {
        @Cache
        final class Default {

        }
        return Default.class.getAnnotation(Cache.class);
    }

    private WebScript getDefaultWebScriptAnnotation() {
        @WebScript
        final class Default {

        }
        return Default.class.getAnnotation(WebScript.class);
    }

    /* Dependencies */

    @Override
    public void setBeanFactory(@NotNull BeanFactory beanFactory) {
        if (beanFactory == null) {
            throw new IllegalArgumentException("beanFactory is null");
        }
        Assert.isInstanceOf(ConfigurableListableBeanFactory.class, beanFactory,
                "BeanFactory is not of type ConfigurableListableBeanFactory.");
        this.beanFactory = (ConfigurableListableBeanFactory) beanFactory;
    }
}

File: annotations-runtime/src/main/java/com/github/dynamicextensionsalfresco/webscripts/AnnotationWebScriptBuilder.kt
package com.github.dynamicextensionsalfresco.webscripts

import com.github.dynamicextensionsalfresco.util.hasText
import com.github.dynamicextensionsalfresco.webscripts.annotations.*
import java.lang.reflect.Method
import java.util.ArrayList
import java.util.Arrays
import java.util.HashSet
import java.util.LinkedHashSet

import com.github.dynamicextensionsalfresco.webscripts.arguments.HandlerMethodArgumentsResolver

import org.springframework.beans.factory.BeanFactory
import org.springframework.beans.factory.BeanFactoryAware
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory
import org.springframework.core.annotation.AnnotationUtils
import org.springframework.extensions.webscripts.Description
import org.springframework.extensions.webscripts.Description.RequiredAuthentication
import org.springframework.extensions.webscripts.Description.RequiredTransaction
import org.springframework.extensions.webscripts.Description.TransactionCapability
import org.springframework.extensions.webscripts.DescriptionImpl
import org.springframework.extensions.webscripts.TransactionParameters
import org.springframework.util.Assert
import org.springframework.util.ClassUtils
import org.springframework.util.ReflectionUtils
import org.springframework.util.StringUtils
import kotlin.properties.Delegates

/**
 * Creates [AnnotationWebScript] instances from beans defined in a [BeanFactory].

 * @author Laurens Fridael
 * @author Laurent Van der Linden
 */

public class AnnotationWebScriptBuilder constructor(
	) : BeanFactoryAware {


    @Autowired
    private var handlerMethodArgumentsResolver: HandlerMethodArgumentsResolver? = null
    public fun setHandlerMethodArgumentsResolver(value:HandlerMethodArgumentsResolver){
        this.handlerMethodArgumentsResolver = value;
    }
    /* Dependencies */

    protected var beanFactory: ConfigurableListableBeanFactory? = null

    private val trailingSlashExpression = "/$".toRegex()
    private val leadingSlashExpression = "^/".toRegex()

    /* Main operations */

    /**
     * Creates [AnnotationWebScript]s from a given named bean by scanning methods annotated with [Uri].

     * @param beanName
     * *
     * @return The [AnnotationWebScript] or null if the implementation does not consider the bean to be a handler
     * *         for an [AnnotationWebScript].
     */
    public fun createWebScripts(beanName: String): List<org.springframework.extensions.webscripts.WebScript> {
        Assert.hasText(beanName, "Bean name cannot be empty.")

        val beanFactory = beanFactory
        val beanType = beanFactory!!.getType(beanName) ?: return emptyList()
        val webScriptAnnotation = beanFactory.findAnnotationOnBean<WebScript>(beanName, WebScript::class.java) ?: getDefaultWebScriptAnnotation()
        val baseUri = webScriptAnnotation.baseUri
        if (StringUtils.hasText(baseUri) && baseUri.startsWith("/") == false) {
            throw RuntimeException("@WebScript baseUri for class '$beanType' does not start with a slash: '$baseUri'")
        }

        val handlerMethods = HandlerMethods()
        ReflectionUtils.doWithMethods(beanType) { method: Method ->
            val before = AnnotationUtils.findAnnotation<Before>(method, Before::class.java)
            if (before != null) {
                if (AnnotationUtils.findAnnotation<Attribute>(method, Attribute::class.java) != null || AnnotationUtils.findAnnotation<Uri>(method, Uri::class.java) != null) {
                    throw RuntimeException("Cannot combine @Before, @Attribute and @Uri on a single method. Method: ${ClassUtils.getQualifiedMethodName(method)}")
                }
                handlerMethods.beforeMethods.add(method)
            }
        }
        ReflectionUtils.doWithMethods(beanType) { method: Method ->
            val attribute = AnnotationUtils.findAnnotation<Attribute>(method, Attribute::class.java)
            if (attribute != null) {
                if (AnnotationUtils.findAnnotation<Before>(method, Before::class.java) != null || AnnotationUtils.findAnnotation<Uri>(method, Uri::class.java) != null) {
                    throw RuntimeException(("Cannot combine @Before, @Attribute and @Uri on a single method. Method: ${ClassUtils.getQualifiedMethodName(method)}"))
                }
                if (method.returnType == Void.TYPE) {
                    throw RuntimeException("@Attribute methods cannot have a void return type.")
                }
                handlerMethods.attributeMethods.add(method)
            }
        }
        ReflectionUtils.doWithMethods(beanType) { method: Method ->
            val exceptionHandler = AnnotationUtils.findAnnotation<ExceptionHandler>(method, ExceptionHandler::class.java)
            if (exceptionHandler != null) {
                if (AnnotationUtils.findAnnotation<Attribute>(method, Attribute::class.java) != null || AnnotationUtils.findAnnotation<Before>(method, Before::class.java) != null || AnnotationUtils.findAnnotation<Uri>(method, Uri::class.java) != null) {
                    throw RuntimeException("Cannot combine @Before, @Attribute @ExceptionHandler or @Uri on a single method. Method: ${ClassUtils.getQualifiedMethodName(method)}")
                }
                handlerMethods.exceptionHandlerMethods.add(ExceptionHandlerMethod(exceptionHandler, method))
            }
        }

        val webScripts = ArrayList<org.springframework.extensions.webscripts.WebScript>()
        ReflectionUtils.doWithMethods(beanType) { method: Method ->
            val uri = AnnotationUtils.findAnnotation<Uri>(method, Uri::class.java)
            if (uri != null) {
                val webScript = createWebScript(beanName, webScriptAnnotation, uri, handlerMethods.createForUriMethod(method))
                webScripts.add(webScript)
            }
        }

        val ids = HashSet<String>()
        for (webScript in webScripts) {
            val webscriptId = webScript.description.id
            val notContained = ids.add(webscriptId)
            if (!notContained) {
                throw IllegalStateException("Duplicate Web Script ID \"" + webscriptId + "\" Make sure handler methods of annotation-based Web Scripts have unique names.")
            }
        }

        return webScripts
    }

    /* Utility operations */

    protected fun createWebScript(beanName: String, webScript: WebScript, uri: Uri, handlerMethods: HandlerMethods): AnnotationWebScript {
        val description = DescriptionImpl()
        if (webScript.defaultFormat.hasText()) {
            description.defaultFormat = webScript.defaultFormat
        }
        val baseUri = webScript.baseUri
        handleHandlerMethodAnnotation(uri, handlerMethods.uriMethod, description, baseUri)
        handleTypeAnnotations(beanName, webScript, description)
        val id = "%s.%s.%s".format(generateId(beanName), handlerMethods.uriMethod.name, description.method.toLowerCase())
        description.id = id
        val handler = beanFactory!!.getBean(beanName)
        description.store = DummyStore()
        return createWebScript(description, handler, handlerMethods)
    }

    protected fun createWebScript(description: Description, handler: Any, handlerMethods: HandlerMethods): AnnotationWebScript {
        return AnnotationWebScript(description, handler, handlerMethods, handlerMethodArgumentsResolver)
    }

    protected fun handleHandlerMethodAnnotation(uri: Uri, method: Method, description: DescriptionImpl, baseUri: String) {
        Assert.notNull(uri, "Uri cannot be null.")
        Assert.notNull(method, "HttpMethod cannot be null.")
        Assert.notNull(description, "Description cannot be null.")

        val uris: Array<String>
        if (uri.value.size > 0) {
            uris = uri.value.map { "${baseUri.replace(trailingSlashExpression, "")}/${it.replace(leadingSlashExpression, "")}" }.toTypedArray()
        } else if (StringUtils.hasText(baseUri)) {
            uris = arrayOf(baseUri.replace(trailingSlashExpression, ""))
        } else {
            throw RuntimeException(
                    "No value specified for @Uri on method '%s' and no base URI found for @WebScript on class."
                            .format(ClassUtils.getQualifiedMethodName(method))
            )
        }
        description.setUris(uris)
        /*
		 * For the sake of consistency we translate the HTTP method from the HttpMethod enum. This also shields us from
		 * changes in the HttpMethod enum names.
		 */
        description.method = when (uri.method) {
            HttpMethod.GET -> "GET"
            HttpMethod.POST -> "POST"
            HttpMethod.PUT -> "PUT"
            HttpMethod.DELETE -> "DELETE"
            HttpMethod.OPTIONS -> "OPTIONS"
        }
        /*
		 * Idem dito for FormatStyle.
		 */
        description.formatStyle = when (uri.formatStyle) {
            FormatStyle.ANY -> Description.FormatStyle.any
            FormatStyle.ARGUMENT -> Description.FormatStyle.argument
            FormatStyle.EXTENSION -> Description.FormatStyle.extension
        }
        if (uri.defaultFormat.hasText()) {
            description.defaultFormat = uri.defaultFormat
        }
        description.multipartProcessing = uri.multipartProcessing

        val methodAuthentication = method.getAnnotation<Authentication>(Authentication::class.java)
        if (methodAuthentication != null) {
            handleAuthenticationAnnotation(methodAuthentication, description)
        }

        val methodTransaction = method.getAnnotation<Transaction>(Transaction::class.java)
        if (methodTransaction != null) {
            handleTransactionAnnotation(methodTransaction, description)
        }
    }

    protected fun handleTypeAnnotations(beanName: String, webScript: WebScript, description: DescriptionImpl) {
        handleWebScriptAnnotation(webScript, beanName, description)

        if (description.requiredAuthentication == null) {
            var authentication = beanFactory!!.findAnnotationOnBean<Authentication>(beanName, Authentication::class.java)
                ?: getDefaultAuthenticationAnnotation()
            handleAuthenticationAnnotation(authentication, description)
        }

        if (description.requiredTransactionParameters == null) {
            var transaction = beanFactory!!.findAnnotationOnBean<Transaction>(beanName, Transaction::class.java)
            if (transaction == null) {
                if (description.method.equals("GET")) {
                    transaction = getDefaultReadonlyTransactionAnnotation()
                } else {
                    transaction = getDefaultReadWriteTransactionAnnotation()
                }
            }
            handleTransactionAnnotation(transaction, description)
        }

        val cache = beanFactory!!.findAnnotationOnBean<Cache>(beanName, Cache::class.java) ?: getDefaultCacheAnnotation()
        handleCacheAnnotation(cache, beanName, description)

        description.descPath = ""
    }

    protected fun handleWebScriptAnnotation(webScript: WebScript, beanName: String, description: DescriptionImpl) {
        Assert.notNull(webScript, "Annotation cannot be null.")
        Assert.hasText(beanName, "Bean name cannot be empty.")
        Assert.notNull(description, "Description cannot be null.")
        Assert.hasText(description.method, "Description method is not specified.")

        if (webScript.value.hasText()) {
            description.shortName = webScript.value
        } else {
            description.shortName = generateShortName(beanName)
        }
        if (webScript.description.hasText()) {
            description.description = webScript.description
        } else {
            description.description = "Annotation-based WebScript for class %s".format(beanFactory!!.getType(beanName).name)
        }
        if (webScript.families.size > 0) {
            description.familys = LinkedHashSet(Arrays.asList(*webScript.families))
        }
        description.lifecycle = when (webScript.lifecycle) {
            Lifecycle.NONE -> Description.Lifecycle.none
            Lifecycle.DRAFT -> Description.Lifecycle.draft
            Lifecycle.DRAFT_PUBLIC_API -> Description.Lifecycle.draft_public_api
            Lifecycle.DEPRECATED -> Description.Lifecycle.deprecated
            Lifecycle.INTERNAL -> Description.Lifecycle.internal
            Lifecycle.PUBLIC_API -> Description.Lifecycle.public_api
            Lifecycle.SAMPLE -> Description.Lifecycle.sample
        }
    }

    protected fun handleAuthenticationAnnotation(authentication: Authentication, description: DescriptionImpl) {
        Assert.notNull(authentication, "Annotation cannot be null.")
        Assert.notNull(description, "Description cannot be null.")
        if (authentication.runAs.hasText()) {
            description.runAs = authentication.runAs
        }
        description.requiredAuthentication = when (authentication.value) {
            AuthenticationType.NONE -> RequiredAuthentication.none
            AuthenticationType.GUEST -> RequiredAuthentication.guest
            AuthenticationType.USER -> RequiredAuthentication.user
            AuthenticationType.ADMIN -> RequiredAuthentication.admin
        }
    }

    protected fun handleTransactionAnnotation(transaction: Transaction, description: DescriptionImpl) {
        Assert.notNull(transaction, "Annotation cannot be null.")
        Assert.notNull(description, "Description cannot be null.")

        val transactionParameters = TransactionParameters()
        transactionParameters.required = when (transaction.value) {
            TransactionType.NONE -> RequiredTransaction.none
            TransactionType.REQUIRED -> RequiredTransaction.required
            TransactionType.REQUIRES_NEW -> RequiredTransaction.requiresnew
        }
        if (transaction.readOnly) {
            transactionParameters.capability = TransactionCapability.readonly
        } else {
            transactionParameters.capability = TransactionCapability.readwrite
        }
        transactionParameters.bufferSize = transaction.bufferSize
        description.requiredTransactionParameters = transactionParameters
    }

    protected fun handleCacheAnnotation(cache: Cache, beanName: String, description: DescriptionImpl) {
        Assert.notNull(cache, "Annotation cannot be null.")
        Assert.hasText(beanName, "Bean name cannot be empty.")
        Assert.notNull(description, "Description cannot be null.")

        val requiredCache = org.springframework.extensions.webscripts.Cache()
        requiredCache.neverCache = cache.neverCache
        requiredCache.isPublic = cache.isPublic
        requiredCache.mustRevalidate = cache.mustRevalidate
        description.requiredCache = requiredCache
    }

    protected fun generateId(beanName: String): String {
        Assert.hasText(beanName, "Bean name cannot be empty")
        val clazz = beanFactory!!.getType(beanName)
        return clazz.name
    }

    protected fun generateShortName(beanName: String): String {
        Assert.hasText(beanName, "Bean name cannot be empty")
        val clazz = beanFactory!!.getType(beanName)
        return ClassUtils.getShortName(clazz)
    }

    /*
	 * These methods use local classes to obtain annotations with default settings.
	 */
    private fun getDefaultAuthenticationAnnotation(): Authentication {
        @Authentication
        class Default
        return Default::class.java.getAnnotation(Authentication::class.java)
    }

    private fun getDefaultReadWriteTransactionAnnotation(): Transaction {
        @Transaction
        class Default
        return Default::class.java.getAnnotation(Transaction::class.java)
    }

    private fun getDefaultReadonlyTransactionAnnotation(): Transaction {
        @Transaction(readOnly = true)
        class Default
        return Default::class.java.getAnnotation(Transaction::class.java)
    }

    private fun getDefaultCacheAnnotation(): Cache {
        @Cache
        class Default
        return Default::class.java.getAnnotation(Cache::class.java)
    }

    private fun getDefaultWebScriptAnnotation(): WebScript {
        @WebScript
        class Default
        return Default::class.java.getAnnotation(WebScript::class.java)
    }

    /* Dependencies */

    override fun setBeanFactory(beanFactory: BeanFactory) {
        Assert.isInstanceOf(ConfigurableListableBeanFactory::class.java, beanFactory, "BeanFactory is not of type ConfigurableListableBeanFactory.")
        this.beanFactory = beanFactory as ConfigurableListableBeanFactory
    }
}
